### PR TITLE
add stop, isTerminated, isShutdown

### PR DESCRIPTION
**Pull request by thespags**

### CODE REVIEW

File: executor-core/src/main/java/net/spals/appbuilder/executor/core/DefaultExecutorServiceFactory.java [hunk 1]
```diff
@@ -105,24 +100,46 @@ public ScheduledExecutorService createScheduledThreadPool(
         return scheduledExecutorService;
     }
 
-    ExecutorService decorateExecutorService(final ExecutorService delegate) {
+    @Override
+    public Optional<Boolean> stop(final Key key) {
+        Objects.requireNonNull(key);
+        final Optional<ExecutorService> executoreService = Optional.ofNullable(executorServices.get(key));
+        if (executoreService.isPresent()) {
```
Comment (thespags) [inline executor-core/src/main/java/net/spals/appbuilder/executor/core/DefaultExecutorServiceFactory.java]: I feel like I could be slicking with my options here to have no if statements but I couldn't figure it out.

File: executor-core/src/main/java/net/spals/appbuilder/executor/core/ExecutorServiceFactory.java [hunk 2]
```diff
@@ -66,16 +67,48 @@ ScheduledExecutorService createScheduledThreadPool(
...
+     * Stops the executor of the given key. If the key does exist, then attempts to shutdown the executor.
+     * If the key does NOT exist, then no-ops and returns false.
+     * <p>
+     * Performs stop via the recommend oracle best practice <a href='https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ExecutorService.html'>here</a>
+     * </p>
+     *
+     * @param key key of the executor
+     * @return true if the key exists and the executor was terminated, false if it exists but already terminated, otherwise empty
+     */
+    Optional<Boolean> stop(Key key);
```
Comment (thespags) [inline executor-core/src/main/java/net/spals/appbuilder/executor/core/ExecutorServiceFactory.java]: I like to work on my javadocing

File: executor-core/src/main/java/net/spals/appbuilder/executor/core/DefaultExecutorServiceFactory.java [hunk 3]
```diff
@@ -134,7 +151,7 @@ Logger getExecutorServiceLogger(final Key key) {
     @PreDestroy
     public synchronized void stop() {
```
Comment (timkral) [inline executor-core/src/main/java/net/spals/appbuilder/executor/core/DefaultExecutorServiceFactory.java]: More clear to name this `stopAll`. Similar to what you did in the tests.

File: executor-core/src/main/java/net/spals/appbuilder/executor/core/DefaultExecutorServiceFactory.java (hunk 1 above)
Comment (timkral) [reply executor-core/src/main/java/net/spals/appbuilder/executor/core/DefaultExecutorServiceFactory.java]: 1. Have `stopExecutorService` return a boolean
2. `Optional.ofNullable(executorServices.get(key)).map(executorService -> stopExecutorService(key, executorService);`

File: executor-core/src/main/java/net/spals/appbuilder/executor/core/DefaultExecutorServiceFactory.java [hunk 4]
```diff
@@ -105,24 +100,46 @@ public ScheduledExecutorService createScheduledThreadPool(
...
+            }
+            executoreService.ifPresent(x -> stopExecutorService(key, x));
+            return executoreService.map(x -> true);
+        }
+        return Optional.empty();
+    }
+
+    @Override
+    public Optional<Boolean> isTerminated(final Key key) {
+        Objects.requireNonNull(key);
```
Comment (timkral) [inline executor-core/src/main/java/net/spals/appbuilder/executor/core/DefaultExecutorServiceFactory.java]: You've used the `@Nonnull` annotation in other places so why the change?

Comment (thespags) [reply executor-core/src/main/java/net/spals/appbuilder/executor/core/DefaultExecutorServiceFactory.java]: I forgot. Good catch.

**Review by timkral — APPROVED** [review]
